lib/sda/host.go

```diff
@@ -368,77 +376,46 @@ func (h *Host) processMessages() {
 				h.checkPendingTreeMarshal(&il)
 			}
 			dbg.Lvl4("Received new entityList")
-		case CosiRequestMessage:
-			// Standalone CoSi hack!
-			// A client wants to sign something
-			cr := data.Msg.(CosiRequest)
-			h.handleCosiRequest(data.Entity, &cr)
 		default:
-			dbg.Error("Didn't recognize message", data.MsgType)
+			err := h.ProcessUnknownMessage(&data)
+			if err != nil {
+				dbg.Error(err, data.MsgType)
+			}
 		}
 		if err != nil {
 			dbg.Error("Sending error:", err)
 		}
 	}
 }
 
-// handleCosiRequest will
-// * register the entitylist
-// * create a flat tree out of it with him being at the root
-// * launch a CoSi protocol
-// * wait for it to finish and send back the response to the client
-func (h *Host) handleCosiRequest(client *network.Entity, cr *CosiRequest) {
-	dbg.Lvl2(h.workingAddress, "Received client CosiRequest from", client.First())
-	idx, e := cr.EntityList.Search(h.Entity.Id)
-	if e == nil {
-		dbg.Error("Received CoSiRequest without being included in the Entitylist")
-		return
-	}
-	if idx != 0 {
-		// replace the first entity by this host's entity
-		tmp := cr.EntityList.List[0]
-		cr.EntityList.List[0] = e
-		cr.EntityList.List[idx] = tmp
-	}
-	// register & create the tree
-	h.overlay.RegisterEntityList(cr.EntityList)
-	// for the moment let's just stick to a very simple binary tree
-	tree := cr.EntityList.GenerateBinaryTree()
-	h.overlay.RegisterTree(tree)
-
-	// run the CoSi protocol
-	node, err := h.overlay.CreateNewNodeName("CoSi", tree)
-	if err != nil {
-		dbg.Error("Error creating tree upon client request:", err)
-		return
-	}
-	node.ProtocolInstance().SigningMessage(cr.Message)
-	hash := h.Suite().Hash()
-	if _, err := hash.Write(cr.Message); err != nil {
-		dbg.Error("Couldn't hash message:", err)
+// RegisterExternalMessage takes a message-type and registers a function
+// that will be called if this message arrives
+func (h *Host) RegisterExternalMessage(mt network.ProtocolMessage, f ExternalAPI) network.MessageTypeID {
+	mtID := network.RegisterMessageType(mt)
+	h.externalMsgs[mtID] = f
+	return mtID
+}
+
+// ProcessUnknownMessage takes a message and looks it up in
+// the unknown messages
+func (h *Host) ProcessUnknownMessage(m *network.Message) error {
```

Comment:
Isn't this name a little misleading? It processes if the message is registered (know) in the externalMsgs map.